MySQL & PSQL base for MicroMDM

Lepidopteron opened the pull request:
By Default BoltDB is used. In order to switch to another rdbms use the flag `rdbms`:
```
flRdbms = flagset.String("rdbms", env.String("MICROMDM_RDBMS", "boltdb"), "Select RDBMS ['boltdb', 'mysql', 'postgres']")
```



E.g.
```
sudo ./micromdm serve \
    -config-path $(echo $(pwd)) \
    -api-key secret \
    -server-url https://localhost/ \
    -rdbms postgres \
    -rdbms-username micromdm \
    -rdbms-password micromdm \
    -rdbms-database micromdm_test \
    -rdbms-host 127.0.0.1 \
    -rdbms-port 5432 \
    -http-debug true \
    -http-addr http://localhost/ \
    -tls=false
```

knightsc commented on cmd/micromdm/serve.go:
```diff
@@ -144,12 +159,15 @@ func serve(args []string) error {
 
-	devWorker := device.NewWorker(devDB, sm.PubClient, logger)
+	//devWorker := device.NewWorker(devDB, sm.PubClient, logger)
+	devWorker := device.NewWorker(devMysqlDB, sm.PubClient, logger)
```
It would be best to not leave commented out lines in the code base. Additionally methods like `device.NewWorker` or `device.New` down below on line 237 take interfaces as input parameters for the DB argument. So you don't necessarily have to rename the `devDB` variable above to `devMysqlDB`. As long as your `devicemysql.NewDB` call returns an object that conforms to the interface needed you should be able to just say

```
devDB, err := devicemysql.NewDB(sm.MysqlDB)
```

And then the rest of the code should just work.

It would be preferred I think though to keep boltDB as the default and have the config setting dynamically choose the correct DB implementation. Just a couple initial thoughts.

Lepidopteron replied:
@knightsc Thank you for this review, left both lines in place so a developer could easily choose, which lines to enable/disable.

Even better would be a dynamic path based on given input flags.

I will modify the code base these days, so it will take into account, whether mysql settings are provided and switch to mysql or whether not and stick to the former approach of a boltdb.

groob commented:
Hi @Lepidopteron 

This is an unexpected, but welcome PR :) I skimmed through the entirety of it and while it's huge, I don't think it will be a problem to merge. I do have some thoughts though: 

1. Keep boltdb as the default. Make the mysql db available as a configuration option at startup. This is non-negotiable as switching the backend will prevent existing users from updating cleanly. Bolt is also a great option for a lot of users who do not care about HA support and are looking for options vs ProfileManager. 
New users can choose mysql vs bolt,  while existing users can find a path to upgrade if they want to. 

2) I noticed some of the migrations are in the code. Let's move all the schema migrations to the migration folder and eliminate them from the code.

Lepidopteron commented:
@groob I finally had some time to merge all the recent changes to this MySQL Branch.

I am not too confident with the "double" references for either mysql or boltdb in `server/server.go` a general datastore or factory would be cooler.
Especially, if one takes into account to support psql in the future.

Lepidopteron commented:
Setting up new iPads not working anymore after merges. Will fix this first.

Lepidopteron commented:
By default bolt as db is used.
If mysql parameters are provided. MySQL is setup.

**Testing Requirements**
Make sure MySQL is locally available / on the machine you want to run the service.
Also "goose" is used for easy setup
```
brew install goose
brew install mysql
brew postinstall mysql
```

**Prepare & Build**
For testing a sample setup:

```
make db-mysql-migrate-test
make build
```

**Run**
Start the service as followed, we use a random bolt db here, so you can still see what is stored in there. The db is stored in `./assets`.
If you start the service later again, the service should still work, as it does not rely on the bolt db, but the data in your Mysql Database instead

```
sudo ./build/darwin/micromdm serve \
    -config-path $(echo $(pwd)/assets/$(uuidgen)/) \
    -api-key secret \
    -scep-client-validity 10000 \
    -server-url https://my.mdm.com/ \
    -command-webhook-url http://127.0.0.1:5000/webhook \
    -mysql-username micromdm \
    -mysql-password micromdm \
    -mysql-database micromdm_test \
    -mysql-host 127.0.0.1 \
    -mysql-port 3306
```

webbmorris commented:
Hi @Lepidopteron - thank you for this enhancement, it's a terrific adjustment!  Question for you - do your changes include an automated migration from Bolt to MySQL of the existing configuration and device data?  I've run your changes up against a test environment here and didn't see it occur, but figured I should check my expectations before going too far down the path.

Lepidopteron commented:
> Hi @Lepidopteron - thank you for this enhancement, it's a terrific adjustment! Question for you - do your changes include an automated migration from Bolt to MySQL of the existing configuration and device data? I've run your changes up against a test environment here and didn't see it occur, but figured I should check my expectations before going too far down the path.

Thank you @webbmorris, I really appreciate your feedback!

So far, I did not implement a migration. But it is a great idea to provide such a functionality.
I think, a BoltDB export in general would be a good idea, especially with respect to the Postgres support as well.

webbmorris commented:
> > Hi @Lepidopteron - thank you for this enhancement, it's a terrific adjustment! Question for you - do your changes include an automated migration from Bolt to MySQL of the existing configuration and device data? I've run your changes up against a test environment here and didn't see it occur, but figured I should check my expectations before going too far down the path.
> 
> Thank you @webbmorris, I really appreciate your feedback!
> 
> So far, I did not implement a migration. But it is a great idea to provide such a functionality.
> I think, a BoltDB export in general would be a good idea, especially with respect to the Postgres support as well.

Thank you @Lepidopteron for the quick response - we will see if we can't support the project by implementing.

MobileDan commented:
I plan to integrate my own microservice datasource. In preparation, I have been reviewing and testing the code in this pull request for a few days. It helped me to see a working example of a second datasource integration - thank you @Lepidopteron.

I have fixed a few issues during my testing and would like to contribute back. Please advise me on the best way to propose my changes. I have used Git for a long time but am new to pull requests.

I effectively want to do a pull request on a pull request. I planned to branch the pull request branch (pr/558) and then do a pull request of that branch. Let me know if that is the best route.

groob commented:
Hi @MobileDan 

I suggest you contribute to @Lepidopteron's fork. While adding a real database is one of the goals I have in mind, it will likely look different than the changes here.

I personally have not reviewed the work Tobias did, and while I might in the future there's no progress happening on my side to help get this pull request merged into the main branch, so development in a fork is best.

Lepidopteron commented:
Hey @MobileDan 

That’s cool, looking forward to your insights.
As of now we had this data structure in use for our production system.
If there are no breaking changes but only improvements, I think committing to this PR is just fine.

If there are major changes and data migrations, I would ask you to go ahead with the approach you mentioned.
Take a branch of this branch and later on merge it back.
I would have a look at the commit from you anyway :-)

Thank you for your commitment. I really appreciate that people do benefit from this work!

Lepidopteron commented:
* CircleCI enhanced to cover pg and mysql tests
* Merged latest changes of micromdm/micromdm

Lepidopteron commented:
@groob, @jessepeterson Would appreciate your feedback on this.
This branch does support psql AND mysql for micromdm.

In order to run on a different rdbms than bolt is configurable via the flag `rdbms` (default `boltdb`, `mysql`, `postgres`)

```
sudo ./micromdm serve \
    -config-path $(echo $(pwd)) 
    -api-key secret \
    -scep-client-validity 10000 \
    -server-url https://localhost/ \
    -rdbms mysql \
    -rdbms-username root \
    -rdbms-password root \
    -rdbms-database micromdm \
    -rdbms-host 127.0.0.1 \
    -rdbms-port 3306
```

groob commented:
I took a quick look. Before going further I think you should clean up the pull request and remove things like kubernetes helm charts and private data which is not part of the proposed changes. 

it will be easier to review and I also think some files were committed by mistake.

Lepidopteron commented:
> I took a quick look. Before going further I think you should clean up the pull request and remove things like kubernetes helm charts and private data which is not part of the proposed changes.
> 
> it will be easier to review and I also think some files were committed by mistake.

Thank you @groob I removed some added folders, which might not be required for everyone.
On `platform/config/mysql/` and `platform/config/pq/` I added a random `dep_token.p7m` and `dep_public_key.pem` for testing. If you consider this as 'not appropriate' I will remove em as well from the history 🙂

groob commented:
I checked out the implementation. It seems fine, but I'm hesitant to merge it in. In my opinion this kind of thing would be best if it lived as a fork. The support burden of accepting these changes is more than I'm willing to provide myself.

crashdump commented on Makefile:
```diff
@@ -146,3 +155,41 @@ define goose_up
 	cd ./pg/migrations && goose postgres "host=${PG_HOST} port=5432 user=micromdm dbname=$(1) password=micromdm sslmode=disable" up
```
`sslmode` should be configurable.

crashdump commented:
We're looking to use MicroMDM for a fairly large estate, which would very likely require horizontal scaling and an external database. I'm very interested by this PR, hopefuly it gets merged.

grahamgilbert commented:
Whilst a real database is desirable, I can tell you that micro will serve five figures of devices from a single 32gb ram instance pretty easily.

crashdump commented:
Thank you @grahamgilbert this is good to know. Is there a way to achieve a highly-available deployment with the bolt backend?

crashdump commented on server/server.go:
```diff
@@ -195,22 +325,74 @@ func (c *Server) setupCommandQueue(logger log.Logger) error {
+		db, err := dbutil.OpenDBX(
+			"postgres",
+			//"host=localhost port=5432 user=micromdm dbname=micromdm_test password=micromdm sslmode=disable",
+			"host="+c.RdbmsHost+" port="+c.RdbmsPort+" user="+c.RdbmsUsername+" dbname="+c.RdbmsDatabase+" password="+c.RdbmsPassword+" sslmode=disable",
```
`sslmode` should be exposed so it can be set to _require_ or _disabled_ as needed.
